### ISSUE TITLE
mongoose connection issues

### ISSUE THREAD

**Issue by baldur** [polygon-city/polygon-city-app#110]
@robhawkes can you add some logging with details around mongoose connection glitches? I am having a problem after you upgraded mongoose / or after I made a external mongodb host where connection drops and the loadbalancer pulls out the node which knocks dev.polygon.city out.

I found this while doing a very brief research https://github.com/Automattic/mongoose/issues/2656 which suggest that some changes occurred around this. If we can log what is happening on connection error and then instead of rescuing just making the server stop and upstart will kick it up again (which is fine in the short term). Here is a trace of what I am seeing.

```bash
/srv/building/releases/079264a77f999a1951a1dc582e8f249ca343cf69/node_modules/mongoose/node_modules/mongodb/lib/utils.js:97
    process.nextTick(function() { throw err; });
                                        ^
Error
    at Error.MongoError (/srv/building/releases/079264a77f999a1951a1dc582e8f249ca343cf69/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/error.js:13:17)
    at Function.MongoError.create (/srv/building/releases/079264a77f999a1951a1dc582e8f249ca343cf69/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/error.js:31:
11)
    at /srv/building/releases/079264a77f999a1951a1dc582e8f249ca343cf69/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:768:66
    at Callbacks.emit (/srv/building/releases/079264a77f999a1951a1dc582e8f249ca343cf69/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:84:
3)
    at null.messageHandler (/srv/building/releases/079264a77f999a1951a1dc582e8f249ca343cf69/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.j
s:218:23)
    at Socket.<anonymous> (/srv/building/releases/079264a77f999a1951a1dc582e8f249ca343cf69/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/connection/connectio
n.js:259:22)
    at Socket.EventEmitter.emit (events.js:95:17)
    at Socket.<anonymous> (_stream_readable.js:746:14)
    at Socket.EventEmitter.emit (events.js:92:17)
    at emitReadable_ (_stream_readable.js:408:10)

/srv/building/releases/079264a77f999a1951a1dc582e8f249ca343cf69/node_modules/mongoose/node_modules/mongodb/lib/utils.js:97
    process.nextTick(function() { throw err; });
                                        ^
Error
    at Error.MongoError (/srv/building/releases/079264a77f999a1951a1dc582e8f249ca343cf69/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/error.js:13:17)
    at Function.MongoError.create (/srv/building/releases/079264a77f999a1951a1dc582e8f249ca343cf69/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/error.js:31:11)
    at /srv/building/releases/079264a77f999a1951a1dc582e8f249ca343cf69/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:768:66
    at Callbacks.emit (/srv/building/releases/079264a77f999a1951a1dc582e8f249ca343cf69/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:84:3)
    at null.messageHandler (/srv/building/releases/079264a77f999a1951a1dc582e8f249ca343cf69/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:218:23)
    at Socket.<anonymous> (/srv/building/releases/079264a77f999a1951a1dc582e8f249ca343cf69/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/connection/connection.js:259:22)
:
```

**Comment by baldur** [comment]
Hey @robhawkes I took a quick stab at the connection issues https://github.com/polygon-city/building-database/commit/cd6adc2214a7be9c1aa8ff2c438faf02ce37338d there is allot of discussion about this online. This particular patch I lifted from this thread http://stackoverflow.com/questions/16226472/mongoose-autoreconnect-option

Best I could do to repro this was to restart mongo and realize that the node app would find it self in a bad state. I see two action points from this ... make sure the the mongoose connection is setup according to best practices and then try to understand why we are seeing these connection blimps.

**Comment by robhawkes** [comment]
Cool ok, let's see how things go with the new error handling and logging. I'm curious what's causing mongo to crap out every so often as I often leave the database running locally for extended periods of time without issue. Let me know if the handlers aren't getting us back into a good state and I'll see what else we can do.

**Comment by robhawkes** [comment]
How are things going now there's some logging in place? Are the connection issues still occurring?

**Comment by baldur** [comment]
it's less frequent as I made some changes we still need to make sure the application can reconnect in such event like restarting mongodb etc.

**Comment by robhawkes** [comment]
I've done a little bit of testing locally and can't replicate this (even without the new logging and reconnect logic), so I assume it's something specific about the dev environment. We're using the latest Mongoose and MongoDB driver, which should have fixed the issues mentioned in the link you posted.

179f297 removes the reconnect on disconnect as it was getting stuck in a loop and trying to open hundreds of new connections when I forcefully closed Mongo. The auto-reconnect logic should kick in just fine with this removed (which is actually the default behaviour of Mongoose), though I haven't tested this on the dev environment (works fine locally). The forced disconnect on error should help as the auto-reconnect will kick in should the database come back up again.

In short: The app seems to reconnect and recover the database connection on a Mongo failure or forced disconnect. No errors that I can see. There's not much more I can do until we can replicate this or get some more info on an error should one occur (hopefully from the new logging).

**Comment by meetar** [comment]
I've not seen any trouble here recently, assume this is resolved
